File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        fast = head.next
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        second = slow.next
        prev = slow.next = None

        first = head

        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp
        second = prev
        while second:
            temp1, temp2 = first.next, second.next
            first.next = second
            second.next = temp1
            second = temp2
            first = temp1
        
        




        
        

        

File: Data Structures & Algorithms/find-duplicate-integer/submission-1.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        const int n = nums.size();
        if (n == 1) return 1;
        int slow = nums[0], fast = nums[nums[0]];
        while (nums[slow] != nums[fast]) {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }

        slow = 0;
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
};
